almanac.lic

```diff
@@ -2,67 +2,91 @@
-    return if XMLData.room_title.include? 'Carousel Booth'
-    return if XMLData.room_title.include? 'Family Vault'
-    return if hidden?
-    return if checkleft
+  def passive_loop
+    loop do
+      use_almanac if should_use_almanac?
+      pause 20
+    end
+  end
 
+  def use_almanac
     unless @almanac_skills.empty?
       training_skill = @almanac_skills
                        .select { |skill| DRSkill.getxp(skill) < 18 }
                        .sort_by { |skill| [DRSkill.getxp(skill), DRSkill.getrank(skill)] }.first
       return unless training_skill
     end
 
-    pause 1 until pause_all
+    # Pause scripts to prevent interference
+    scripts_to_unpause = []
+    scripts_to_unpause = DRC.smart_pause_all
     waitrt?
-    case bput("get my #{@almanac}", 'You get', 'What were', 'You are already', 'You need a free')
-    when 'What were'
-      echo('Almanac not found, exiting.')
-      unpause_all
+
+    # Wait for any last output from paused scripts to resolve
+    # to mitigate things like combat-trainer doing `loot treasure`
+    # and getting a "I could not find what you were referring to" response
+    # (because no critter to loot) before the response to "get my almanac".
+    # In that race condition, we think we didn't get your almanac and
+    # the script exits, but that leaves you with an almanac in your hand
+    # and can cause combat-trainer to hang if it's trying to use a twohanded weapon.
+    pause 1
+    clear
+
+    unless DRCI.get_item_if_not_held?(@almanac) && DRCI.in_hands?(@almanac)
+      DRC.message('Almanac not found, exiting')
+      DRC.unpause_all_list(scripts_to_unpause)
       exit
-    when 'You need a free'
-      unless DRCI.in_hands?(@almanac)
-        unpause_all
-        return
-      end
     end
+
     if training_skill
-      bput("turn #{@almanac} to #{training_skill}", 'You turn', 'You attempt to turn')
+      DRC.bput("turn #{@almanac} to #{training_skill}", 'You turn', 'You attempt to turn')
     end
-    bput("study my #{@almanac}", 'You set about', 'gleaned all the insight you can', 'Study what', 'interrupt your research')
+
+    DRC.bput("study my #{@almanac}", 'You set about', 'gleaned all the insight you can', 'Study what', 'interrupt your research')
     waitrt?
-    bput("stow my #{@almanac}", 'You put', 'Stow what', 'You hold out')
+
+    DRCI.put_away_item?(@almanac)
+
     UserVars.almanac_last_use = Time.now
-    unpause_all
+
+    DRC.unpause_all_list(scripts_to_unpause)
   end
 
-  def passive_loop
-    loop do
-      use_almanac
-      pause 20
-    end
+  def should_use_almanac?
+    !(hidden? || invisible? || almanac_on_cooldown? || hands_full? || running_no_use_scripts? || inside_no_use_room?)
   end
+
+  def running_no_use_scripts?
+    @no_use_scripts.any? { |name| Script.running?(name) }
+  end
+
+  def inside_no_use_room?
+    @no_use_rooms.any? { |room| room === DRRoom.title.to_s()[2..-3] || room == Room.current.id }
```

Comment:
> DRRoom.title.to_s()[2..-3]

This feels very "magical". I realize it's done in a bunch of other scripts, but it looks funny and I can't for the life of me remember why.

> s = "this is a very long string which I am making longer"
> puts s[2..-3]
> is is a very long string which I am making long

This seems to mangle the title a lot...

Reply:
Agree, those nuances should be abstracted.  It's because the room title is enclosed in double brackets [[Some Place]]. I don't know why they weren't stripped out in lich.rbw to begin with

Reply:
I'd love to see that change, but it's not necessary for this PR